# EJERCICIO 2 DATAFRAME TITANIC

## Importamos las librerias necesarias para poder analizar la base de datos.

Usamos la libreria pandas y definimos su nomenclatura a lo largo del notebook como "pd" para poder visualizar y analizar los datos.

In [8]:
import pandas as pd

In [10]:
pd.read_csv("C:/Users/Adm1n/Maestria_EAN/titanic.csv") #Mediante la funcion .read_csv() agregamos la ruta donde se encuentra el archivo para 
                                                       #poder importar la base de datos.

Survived  Pclass                                               Name  \
0           0       3                             Mr. Owen Harris Braund   
1           1       1  Mrs. John Bradley (Florence Briggs Thayer) Cum...   
2           1       3                              Miss. Laina Heikkinen   
3           1       1        Mrs. Jacques Heath (Lily May Peel) Futrelle   
4           0       3                            Mr. William Henry Allen   
..        ...     ...                                                ...   
882         0       2                               Rev. Juozas Montvila   
883         1       1                        Miss. Margaret Edith Graham   
884         0       3                     Miss. Catherine Helen Johnston   
885         1       1                               Mr. Karl Howell Behr   
886         0       3                                 Mr. Patrick Dooley   

        Sex   Age  Siblings/Spouses Aboard  Parents/Children Aboard     Fare  
0      male  22.0                        1                        0   7.2500  
1    female  38.0                        1                        0  71.2833  
2    female  26.0                        0                        0   7.9250  
3    female  35.0                        1                        0  53.1000  
4      male  35.0                        0                        0   8.0500  
..      ...   ...                      ...                      ...      ...  
882    male  27.0                        0                        0  13.0000  
883  female  19.0                        0                        0  30.0000  
884  female   7.0                        1                        2  23.4500  
885    male  26.0                        0                        0  30.0000  
886    male  32.0                        0                        0   7.7500  

[887 rows x 8 columns]

In [12]:
titanic=pd.read_csv("C:/Users/Adm1n/Maestria_EAN/titanic.csv")#Asignamos la lectura de la base de datos a la variable "titanic"

In [14]:
titanic.info() #Determinamos la informacion general de la base de datos titanic

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 887 entries, 0 to 886
Data columns (total 8 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Survived                 887 non-null    int64  
 1   Pclass                   887 non-null    int64  
 2   Name                     887 non-null    object 
 3   Sex                      887 non-null    object 
 4   Age                      887 non-null    float64
 5   Siblings/Spouses Aboard  887 non-null    int64  
 6   Parents/Children Aboard  887 non-null    int64  
 7   Fare                     887 non-null    float64
dtypes: float64(2), int64(4), object(2)
memory usage: 55.6+ KB


El dataframe "titanic" tiene un total de 887 filas y 8 columnas par aun total de 7.096 datos.
Tomando en cuenta que las columnas son nuestras variables y cada fila representa una persona que estuvo en el titanic.
Nuestro tipo de variales son:
1. Numerica:
   Enteros (int)=Survived,Pclass,Siblings/Spuses aboard,Parents/Children aboard.
   
   Decimal (float)=Age,Fare
3. Categorico:
4. Cadena de texto-object (str)=Name,Sex

## Detección de valores nulos o vacios.

Usamos la función .isnull.sum() con esto sabremos que cuantos valores estan vacios en cada una de las variables a lo largo del dataframe.

In [16]:
titanic.isnull().sum() #Ver cantidad de valores nulos o vacios en el dataframe.

Survived                   0
Pclass                     0
Name                       0
Sex                        0
Age                        0
Siblings/Spouses Aboard    0
Parents/Children Aboard    0
Fare                       0
dtype: int64

Como podemos observar ninguna variable posee un valor nulo.

## Detección de filas duplicadas

Aplicamos la función .duplicated() la cual nos dice mediante una confirmación con numeros booleanos si existe o no esta condición. De igual forma, adicionamos la función .sum() para determiar a lo largo de todas las filas la condición de filas duplicadas.

In [18]:
titanic.duplicated().sum() #El dataframe titanic no tiene filas duplicadas.

0

# Limpieza de datos

* Procedemos a cambiar el tipo de dato para la variable "Age" de float a int 

In [20]:
titanic.Age=titanic.Age.astype(int) # Cambiamos el tipo de variable mediante la funcion .astype()

In [22]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 887 entries, 0 to 886
Data columns (total 8 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Survived                 887 non-null    int64  
 1   Pclass                   887 non-null    int64  
 2   Name                     887 non-null    object 
 3   Sex                      887 non-null    object 
 4   Age                      887 non-null    int32  
 5   Siblings/Spouses Aboard  887 non-null    int64  
 6   Parents/Children Aboard  887 non-null    int64  
 7   Fare                     887 non-null    float64
dtypes: float64(1), int32(1), int64(4), object(2)
memory usage: 52.1+ KB


* Creamos una nueva columna para identificar si una persona  "Sobrevivio" o "Fallecio".

In [24]:
titanic['Estado']=['Sobrevivio' if i == 1 else 'Fallecio' for i in titanic['Survived']] # Mediante la función If y For 
                                                                                        # establecemos los criterios y parametros de la nueva fila.

In [28]:
titanic.head() #Observamos las priemras 5 filas del Dataframe.

Survived  Pclass                                               Name  \
0         0       3                             Mr. Owen Harris Braund   
1         1       1  Mrs. John Bradley (Florence Briggs Thayer) Cum...   
2         1       3                              Miss. Laina Heikkinen   
3         1       1        Mrs. Jacques Heath (Lily May Peel) Futrelle   
4         0       3                            Mr. William Henry Allen   

      Sex  Age  Siblings/Spouses Aboard  Parents/Children Aboard     Fare  \
0    male   22                        1                        0   7.2500   
1  female   38                        1                        0  71.2833   
2  female   26                        0                        0   7.9250   
3  female   35                        1                        0  53.1000   
4    male   35                        0                        0   8.0500   

       Estado  
0    Fallecio  
1  Sobrevivio  
2  Sobrevivio  
3  Sobrevivio  
4    Fallecio

* Que pasaria si el titanic no se hubiera hundido y si las personas fueran inmortales. ¿ Cuales serian sus edades actuales ?

In [31]:
titanic['Edad_actual']=titanic.Age+(2024-1942) #Creamos la nueva columna mediante una operación entre variables de tipo int.

In [137]:
titanic.head()

Survived  Pclass                                               Name  \
0         0       3                             Mr. Owen Harris Braund   
1         1       1  Mrs. John Bradley (Florence Briggs Thayer) Cum...   
2         1       3                              Miss. Laina Heikkinen   
3         1       1        Mrs. Jacques Heath (Lily May Peel) Futrelle   
4         0       3                            Mr. William Henry Allen   

      Sex  Age  Siblings/Spouses Aboard  Parents/Children Aboard     Fare  \
0    male   22                        1                        0   7.2500   
1  female   38                        1                        0  71.2833   
2  female   26                        0                        0   7.9250   
3  female   35                        1                        0  53.1000   
4    male   35                        0                        0   8.0500   

       Estado  Edad_actual  
0    Fallecio          104  
1  Sobrevivio          120  
2  Sobrevivio          108  
3  Sobrevivio          117  
4    Fallecio          117

## Clasificación de edad de los pasajeros

In [36]:
titanic['Claf_edad']=['Menor' if i < 18 else 'Adulto' for i in titanic['Age']]
titanic.head()

Survived  Pclass                                               Name  \
0         0       3                             Mr. Owen Harris Braund   
1         1       1  Mrs. John Bradley (Florence Briggs Thayer) Cum...   
2         1       3                              Miss. Laina Heikkinen   
3         1       1        Mrs. Jacques Heath (Lily May Peel) Futrelle   
4         0       3                            Mr. William Henry Allen   

      Sex  Age  Siblings/Spouses Aboard  Parents/Children Aboard     Fare  \
0    male   22                        1                        0   7.2500   
1  female   38                        1                        0  71.2833   
2  female   26                        0                        0   7.9250   
3  female   35                        1                        0  53.1000   
4    male   35                        0                        0   8.0500   

       Estado  Edad_actual Claf_edad  
0    Fallecio          104    Adulto  
1  Sobrevivio          120    Adulto  
2  Sobrevivio          108    Adulto  
3  Sobrevivio          117    Adulto  
4    Fallecio          117    Adulto

## Analisis de datos

In [141]:
# Porcentaje de sobrevivientes
porcentaje_sobr=titanic.Estado.value_counts()/titanic.Estado.count()*100
porcentaje_sobr

Estado
Fallecio      61.443067
Sobrevivio    38.556933
Name: count, dtype: float64

In [157]:
# Porcentaje de pasajares en cada nivel de clase
titanic.Pclass.value_counts()/titanic.Pclass.count()*100

Pclass
3    54.904171
1    24.351747
2    20.744081
Name: count, dtype: float64

In [77]:
# Porcentaje de hombres/mujeres en cada nivel
titanic.groupby('Pclass')['Sex'].value_counts('male')*100

Pclass  Sex   
1       male      56.481481
        female    43.518519
2       male      58.695652
        female    41.304348
3       male      70.431211
        female    29.568789
Name: proportion, dtype: float64

In [192]:
# Porcentaje de pasajeros segun su genero
titanic.Sex.value_counts()/titanic.Sex.count()*100

Sex
male      64.599775
female    35.400225
Name: count, dtype: float64

In [213]:
# Porcentaje de menores de edad
titanic.Claf_edad.value_counts()/titanic.Claf_edad.count()*100

Claf_edad
Adulto    85.343856
Menor     14.656144
Name: count, dtype: float64

In [64]:
# Porcentaje de sobrevivientes y fallecidos por categoria
titanic.groupby('Pclass')['Estado'].value_counts('Sobrevivio')*100 # Mediante la función .groupby agrupamos por el tipo de clase con el criterio si sobrevivio o no 
                                                                       # con el atributo de "Estado"

Pclass  Estado    
1       Sobrevivio    62.962963
        Fallecio      37.037037
2       Fallecio      52.717391
        Sobrevivio    47.282609
3       Fallecio      75.564682
        Sobrevivio    24.435318
Name: proportion, dtype: float64

In [70]:
# Porcentaje de sobrevivientes y fallecidos por genero.
titanic.groupby('Sex')['Estado'].value_counts('Sobrevivio')*100

Sex     Estado    
female  Sobrevivio    74.203822
        Fallecio      25.796178
male    Fallecio      80.977312
        Sobrevivio    19.022688
Name: proportion, dtype: float64

In [74]:
#Porcentaje de sobrevivientes y fallecidos por edad.
titanic.groupby('Claf_edad')['Estado'].value_counts('Sobrevivio')*100

Claf_edad  Estado    
Adulto     Fallecio      63.40819
           Sobrevivio    36.59181
Menor      Fallecio      50.00000
           Sobrevivio    50.00000
Name: proportion, dtype: float64

In [80]:
titanic.describe() # Con la función describe.() podemos observar los resultados estadisticos principales.

Survived      Pclass         Age  Siblings/Spouses Aboard  \
count  887.000000  887.000000  887.000000               887.000000   
mean     0.385569    2.305524   29.455468                 0.525366   
std      0.487004    0.836662   14.129919                 1.104669   
min      0.000000    1.000000    0.000000                 0.000000   
25%      0.000000    2.000000   20.000000                 0.000000   
50%      0.000000    3.000000   28.000000                 0.000000   
75%      1.000000    3.000000   38.000000                 1.000000   
max      1.000000    3.000000   80.000000                 8.000000   

       Parents/Children Aboard       Fare  Edad_actual  
count               887.000000  887.00000   887.000000  
mean                  0.383315   32.30542   111.455468  
std                   0.807466   49.78204    14.129919  
min                   0.000000    0.00000    82.000000  
25%                   0.000000    7.92500   102.000000  
50%                   0.000000   14.45420   110.000000  
75%                   0.000000   31.13750   120.000000  
max                   6.000000  512.32920   162.000000

La edad promedio de los pasajeros que abordaron el titanic era de 29 años en un rango desde 0 años (proceso de embarazo) hasta los 80 años (valor maximo)

## CONCLUSIONES

De acuerdo al previo analisis del dataframe titanic, se pueden concluir los siguientes datos:

* Solo sobrevivio el 38.55% de los pasajeros para un total de 342 pasajeros a bordo del titanic. Para un total de fallecidos de 545 pasajeros.
* Se dio preferencia a salvar a mujeres que a hombres. Es por esto que vemos tan reducido el porcentaje de sobrevivientes en la Clase 3 con un 24.43% ya que el 70.43% eran hombres.
* Para tener mas probabilidad de sobrevivir en el evento del titanic se debia cumplir con los siguientes requisitos:
    Ser menor de edad - ser del genero femenino - pertenecer a la clase 1